src/jsonpatcherproxy.js

```diff
@@ -132,18 +84,87 @@ var JSONPatcherProxy = (function() {
       },
       deleteProperty: function(target, key) {
         if (typeof target[key] !== "undefined") {
-          instance.defaultCallback({
+          this.instance.defaultCallback({
             op: "remove",
             path: (
-              path + "/" + JSONPatcherProxy.escapePathComponent(key.toString())
+              this.path + "/" + JSONPatcherProxy.escapePathComponent(key.toString())
             )
           });
+          const proxyInstance = proxifiedObjectsMap.get(target[key])
+          if(proxyInstance) {
+            disableTrapsForProxy.call(this.instance, proxyInstance);
+          }
         }
         // else {
         return Reflect.deleteProperty(target, key);
       }
-    });
-    return proxy;
+    };
+    /**
+    * Creates an instance of JSONPatcherProxy around your object of interest `root`. 
+    * @param {Object|Array} root - the object you want to wrap
+    * @param {Boolean} showDetachedWarning - whether to log a warning when a detached sub-object is modified @see {@link https://github.com/Palindrom/JSONPatcherProxy#detached-objects} 
+    * @returns {JSONPatcherProxy}
+    * @constructor
+    */
+  function JSONPatcherProxy(root, showDetachedWarning) {
+    
+    // default to true
+    if(typeof showDetachedWarning !== 'boolean')  {
+      showDetachedWarning = true;
+    }
+    
+    this.showDetachedWarning = showDetachedWarning;
+    this.originalObject = root;
+    this.cachedProxy = null;
+    this.isRecording = false;
+    this.userCallback;
+    /**
+     * @memberof JSONPatcherProxy
+     * Restores callback back to the original one provided to `observe`.
+     */
+    this.resume = () => {
+      this.defaultCallback = operation => {
+         this.isRecording && this.patches.push(operation);
+         this.userCallback && this.userCallback(operation);
+      };
+    };
+    /**
+     * @memberof JSONPatcherProxy
+     * Replaces your callback with a noop function.
+     */
+    this.pause = () => {
+      this.defaultCallback = function() {};
+    };
+  }
+  /**
+  * Deep clones your object and returns a new object.
+  */
+  JSONPatcherProxy.deepClone = function(obj) {
+    switch (typeof obj) {
+      case "object":
+        return JSON.parse(JSON.stringify(obj)); //Faster than ES5 clone - http://jsperf.com/deep-cloning-of-objects/5
+      case "undefined":
+        return null; //this is how JSON.stringify behaves for array items
+      default:
+        return obj; //no need to clone primitives
+    }
+  };
+  JSONPatcherProxy.escapePathComponent = function(str) {
+    if (str.indexOf("/") === -1 && str.indexOf("~") === -1) return str;
+    return str.replace(/~/g, "~0").replace(/\//g, "~1");
+  };
+  JSONPatcherProxy.prototype.generateProxyAtPath = function(obj, path) {
+    if (!obj) {
+      return obj;
+    }
+    var instance = this;
+    var trapsInstance = Object.assign({}, traps, {path: path, instance: instance});
```

Comment:
That's right. I thought not defining the functions (`set`, `get` etc..) again and just referencing them would be better. But this would be slower with every issued patch (`this.instance.defaultCallback` is slow).

Moving them inside could be a little slower but for one time only. Faster with every patch.

I moved it inside.